solidity-bindgen-macros/src/abi_gen.rs

```diff
@@ -0,0 +1,159 @@
+use crate::abi_json::{abi_from_json, Abi};
+use ethabi::param_type::{ParamType, Reader};
+use inflector::cases::snakecase::to_snake_case;
+use proc_macro2::{Ident, Span, TokenStream};
+use std::path::Path;
+
+pub fn abi_from_file(path: impl AsRef<Path>, span: Span) -> TokenStream {
+    let name = path
+        .as_ref()
+        .file_stem()
+        .unwrap()
+        .to_str()
+        .unwrap()
+        .to_owned();
+    let bytes = std::fs::read(path).unwrap();
+    let abis = abi_from_json(&bytes);
+    let abi_str = String::from_utf8(bytes).unwrap();
+
+    let struct_name = Ident::new(name.as_str(), span);
+
+    let fns = abis.iter().map(|abi| fn_from_abi(abi, span));
+
+    quote! {
+        pub struct #struct_name {
+            contract: ::solidity_bindgen::internal::ContractWrapper,
+        }
+
+        impl #struct_name {
+            pub fn new(address: ::web3::types::Address, url: &str, event_loop_handle: ::std::sync::Arc<::web3::transports::EventLoopHandle>) -> ::solidity_bindgen::internal::Result<Self> {
+                // Embed ABI into the program
+                let abi = #abi_str;
+
+                // Set up a wrapper so we can make calls
+                let contract = ::solidity_bindgen::internal::ContractWrapper::new(address, abi.as_bytes(), url, event_loop_handle)?;
+                Ok(Self {
+                    contract
+                })
+            }
+
+            #(#fns)*
+        }
+    }
+}
+
+fn param_token_type(param_type: ParamType) -> TokenStream {
+    match param_type {
+        ParamType::Address => quote! { ::web3::types::Address },
+        ParamType::Bytes => quote! { ::std::vec::Vec<u8> },
+        ParamType::Int(size) => match size {
+            256 => quote! { ::solidity_bindgen::internal::Unimplemented },
+            _ => {
+                let name = Ident::new(&format!("i{}", size), Span::call_site());
+                quote! { #name }
+            }
+        },
+        ParamType::Uint(size) => match size {
+            256 => quote! { ::web3::types::U256 },
+            _ => {
+                let name = Ident::new(&format!("u{}", size), Span::call_site());
+                quote! { #name }
+            }
+        },
+        ParamType::Bool => quote! { bool },
+        ParamType::String => quote! { ::std::string::String },
+        ParamType::Array(inner) => {
+            let inner = param_token_type(*inner);
+            quote! { ::std::vec::Vec<#inner> }
+        }
+        ParamType::FixedBytes(len) => quote! { [ u8; #len ] },
+        ParamType::FixedArray(inner, len) => {
+            let inner = param_token_type(*inner);
+            quote! { [#inner; #len] }
+        }
+        ParamType::Tuple(members) => {
+            match members.len() {
+                0 => {
+                    // TODO: Verify that this detokenizes the same
+                    quote! { ::solidity_bindgen::internal::Empty }
```

Comment:
According to https://solidity.readthedocs.io/en/v0.5.3/abi-spec.html#types, empty tuples are possible:
> Types can be combined to a tuple by enclosing them inside parentheses, separated by commas:
>
> (T1,T2,...,Tn): tuple consisting of the types T1, …, Tn, n >= 0
> It is possible to form tuples of tuples, arrays of tuples and so on. It is also possible to form zero-tuples (where n == 0).

So an empty tuple should detokenize to `()`, I think.

Reply:
I agree, but for some reason `()` does not impl `Detokenize`. So, I had to create this type that is effectively the same because of the orphan rules.

Reply:
Ah, ok. The `TODO` confused me. You're saying this is actually solved? Or still needs to be verified?

Reply:
I verified it. I think you're reviewing an old version now because this comment is gone.